### PR TITLE
Make ItemContext available for better diagnositcs

### LINKED ISSUE #62570: Async blocks are leaky about wording wrt. `?` in diagnostics
E.g. consider:

```rust
#![feature(async_await)]

fn foo() {
    let x = async move {
        Err(())?;

        1
    };
}
```

which yields the diagnostic:

```
error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
 --> src/lib.rs:5:9
  |
5 |         Err(())?;
  |         ^^^^^^^^ cannot use the `?` operator in a function that returns `{integer}`
  |
  = help: the trait `std::ops::Try` is not implemented for `{integer}`
  = note: required by `std::ops::Try::from_error`
```

Ostensibly the user might think of `fn foo` as "the function" but it is in fact the `async` block that is "the function" here. That seems like something which would confuse users who are not familiar with the lowering and whatnot (most users).

### PR DESCRIPTION
**Pull request by gilescope**
Fix #62570

### CODE REVIEW

**Comment by gilescope** [comment]
Fixes #62570

File: src/librustc/traits/error_reporting.rs [hunk 1]
```diff
@@ -347,6 +347,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+        let node : Node<'_> = hir.get(hir_id);
+
+        if let hir::Node::Item(
+            hir::Item{kind: hir::ItemKind::Fn(_ ,fn_header ,_ , _), ..}) = &node {
+            if let hir::FnHeader{asyncness: hir::IsAsync::Async, ..} = fn_header {
+                "an async function"
+            } else {
+                "a function"
+            }
+        } else if let hir::Node::Expr(hir::Expr{
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]:
```suggestion
        } else if let hir::Node::Expr(hir::Expr {
```

File: src/librustc/traits/error_reporting.rs [hunk 2]
```diff
@@ -347,6 +347,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+            if let Some(_is_pinned) = generator_movability {
+                "an async closure"
+            } else {
+                "a closure"
+            }
+        } else if let hir::Node::Expr(hir::Expr{..}) = &node {
+            let parent_hid = hir.get_parent_node(hir_id);
+            self.describe_enclosure(parent_hid)
+        } else {
+            bug!("can't describe a {:#?}", node);
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: Why not return `Option<&'static str>` instead?

File: src/librustc/traits/error_reporting.rs [hunk 3]
```diff
@@ -357,6 +383,9 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         let trait_ref = *trait_ref.skip_binder();
 
         let mut flags = vec![];
+        flags.push((sym::item_context,
+            Some(self.describe_enclosure(obligation.cause.body_id).to_owned())));
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: ...and use `if let Some(descr) = self.describe_enclosure<...>` here?

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: I think that with that change you'd fix all of [the errors I'm seeing](https://dev.azure.com/rust-lang/rust/_build/results?buildId=12155).

File: src/librustc/traits/error_reporting.rs [hunk 4]
```diff
@@ -347,6 +347,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+
+        if let hir::Node::Item(
+            hir::Item{kind: hir::ItemKind::Fn(_ ,fn_header ,_ , _), ..}) = &node {
+            if let hir::FnHeader{asyncness: hir::IsAsync::Async, ..} = fn_header {
+                "an async function"
+            } else {
+                "a function"
+            }
+        } else if let hir::Node::Expr(hir::Expr{
+          kind: hir::ExprKind::Closure(_is_move,_,_,_,generator_movability), ..}) = &node {
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]:
```suggestion
          kind: hir::ExprKind::Closure(_is_move, _, _, _, generator_movability), ..}) = &node {
```

File: src/librustc/traits/error_reporting.rs [hunk 5]
```diff
@@ -347,6 +347,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+            if let Some(_is_pinned) = generator_movability {
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]:
```suggestion
            if generator_movability.is_some() {
```

File: src/librustc/traits/error_reporting.rs [hunk 6]
```diff
@@ -347,6 +347,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                "an async closure"
+            } else {
+                "a closure"
+            }
+        } else if let hir::Node::Expr(hir::Expr{..}) = &node {
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]:
```suggestion
        } else if let hir::Node::Expr(hir::Expr { .. }) = &node {
```

File: src/librustc/traits/error_reporting.rs [hunk 7]
```diff
@@ -347,6 +347,53 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         }
     }
 
+    fn describe_generator(&self, body_id: hir::BodyId) -> Option<&'static str> {
+        self.tcx.hir().body(body_id).generator_kind.map(|gen_kind| {
+            match gen_kind {
+                hir::GeneratorKind::Gen => { "a generator" },
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Block) => { "an async block" },
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Fn) => { "an async function" },
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: The braces are unnecessary.

File: src/librustc/traits/error_reporting.rs [hunk 8]
```diff
@@ -347,6 +347,53 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Closure)
+                    => { "an async closure" },
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: Use the braces to put `"an async closure"` on its own line. When using braces the trailing comma is not necessary after the block.

File: src/librustc/traits/error_reporting.rs [hunk 9]
```diff
@@ -347,6 +347,53 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+            }
+        })
+    }
+
+    /// Used to set on_unimplemented's `ItemContext`
+    /// to be the enclosing (async) block/function/closure
+    fn describe_enclosure(&self, hir_id: hir::HirId) -> Option<&'static str> {
+        let hir = &self.tcx.hir();
+        let node : Node<'_> = hir.get(hir_id);
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]:
```suggestion
        let node : Node<'_> = hir.find(hir_id)?;
```

File: src/librustc/traits/error_reporting.rs (hunk 7 above)
Comment (gilescope) [reply src/librustc/traits/error_reporting.rs]: ah perfect, now the last case doesn't tip the 100 char line limit. Thanks!

File: src/librustc/traits/error_reporting.rs (hunk 9 above)
Comment (gilescope) [reply src/librustc/traits/error_reporting.rs]: You can use ? returns with options as well as results?? That's a thing of beauty.

File: src/librustc/traits/error_reporting.rs [hunk 10]
```diff
@@ -347,6 +347,52 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
     }
 
+    fn describe_generator(&self, body_id: hir::BodyId) -> Option<&'static str> {
+        self.tcx.hir().body(body_id).generator_kind.map(|gen_kind| {
+            match gen_kind {
+                hir::GeneratorKind::Gen => { "a generator" },
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Block) => "an async block",
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Fn) => "an async function",
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Closure) => "an async closure",
+            }
```
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: My preferred style here would be:

```
            match gen_kind {
                hir::GeneratorKind::Gen => "a generator",
                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Block) => "an async block",
                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Fn) => "an async function",
                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Closure) => "an async closure",
            }
```

Comment (gilescope) [reply src/librustc/traits/error_reporting.rs]: ah missed removing the curlies around the generator case.

**Pull request merged**